pkg/envoy/xds/server.go

```diff
@@ -295,20 +300,29 @@ func (s *Server) processRequestStream(ctx context.Context, streamLog *logrus.Ent
 			// the last response yet. Ignore every request until it processes
 			// that response and sends a request with that response's nonce.
 			if state.nonce == "" || state.nonce == req.GetResponseNonce() {
-				if state.pendingWatchCancel != nil {
-					// A pending watch exists for this type URL. Cancel it to
-					// start a new watch.
-					requestLog.Debug("canceling pending watch")
-					state.pendingWatchCancel()
-				} else if versionInfo != nil {
-					// If no pending watch exists, then this request is an ACK
-					// for the last response for this resource type.
+				if versionInfo != nil && *versionInfo == state.version {
+					// This request is an ACK.
 					// Notify every observer of the ACK.
 					ackObserver := s.ackObservers[typeURL]
 					if ackObserver != nil {
 						requestLog.Debug("notifying observers of ACK")
 						ackObserver.HandleResourceVersionAck(*versionInfo, req.GetNode(), state.resourceNames, typeURL)
+					} else {
+						requestLog.Debug("ACK received but no observers are waiting for ACKs")
 					}
+				} else if versionInfo != nil {
+					requestLog.Debug("NACK received for version ", state.version, " waiting for a version update before sending again")
```

Comment:
I'd use the Warning level for this one. It will typically indicate a bug in the agent, but one we may be able to recover from.
And use a format string:
```
requestLog.Warningf("NACK received for version %d; waiting for a version update before sending again", state.version)
```